### ISSUE TITLE
Support 17-bit-shifted PCIN input on DSP48E2

### ISSUE THREAD

**Issue by gussmith23** [gussmith23/lakeroad#464]
For Churchroad to implement wide multiplies, we need Lakeroad to support inputting to the PCIN port, which can be internally shifted by 17 bits.

Should wait on #474 and merge that in.

Subtasks:
- [x] #474

**Comment by gussmith23** [comment]
## 2024-09-04

I am working on this in a PR, and I have a test for it. No luck yet. Some things I may want to do:

- Simplify the test so that it's literally just testing whether we can shift an input.
- Finally merge in the commits that clean up the input-specification, so that we can e.g. pass just a C input. Currently we can't do this, but I swear Vishal has commits somewhere that made this a lot cleaner.

**Comment by gussmith23** [comment]
## 2024-11-13

Looking more into this weird issue with the shifter only working on 47 bit or less inputs.

Honestly, it makes me think that getting the counterexamples as a debugging tool would be useful. Though does that make sense? not exactly for synthesis. You get counterexamples if you run a verify query. If you run a synthesis query, you run through a number of verification queries. I guess I could try to get a counterexample from the very last one.

## 2024-11-14

The problem might be that it's always appending the sign bits, ie zero-extension is not possible in the case that the topmost bit is 1. If the topmost bit is 1. So I likely need to just change the spec that I'm trying to synthesize.

Yeah, so if I do an explicit sign-shift in the spec (rather than using >>), we can shift a 48-bit number. I must just be forgetting how to use >> in Verilog. Is it always an unsigned shift? Is there a different operator for arithmetic shift? I assumed that >> with signed inputs would do an arithmetic shift.

In short: the hardware *always* does an arithmetic shift.

Gonna write a few more tests of the internal shifter.

I'm trying to write a test that actually does a multiply-and-add, where the value to be added is shifted. However I'm struggling to write the spec in the way that the hardware needs, which is frustrating---seems like it should be more flexible.

## 2024-11-15

Ok so switching over to cvc5 fixed the timeout for the multiply-and-add-shifted case, so that's good! Makes me wonder if that would have fixed the initial problem too, before I started adding this functionality to Lakeroad. But it needed to be done regardless.

## 2024-11-17

So this PR is probably ready to go (could always use more tests, but it's fine). But we need CI working first, so see #474.
